README.md
# Timeline.js

一个用于动画的时间轴编辑器组件

![Timeline](https://ax.minicg.com/images/timeline-js.jpg)

演示案例:
https://wavef.gitee.io/timeline.js/examples/index.html

项目网站:
http://marcinignac.com/blog/timeline-js/

原作者：
[Marcin Ignac](https://github.com/vorg)

改版：
[WaveF](https://gitee.com/wavef)


## 注意

此版本被我修改后与官方原版存在**较大差异**，需要注意几点：
- `anim()`方法不再是全局对象，需要通过`Timeline.anim()`调用
- 单击关键帧编辑，现在已被改为需要通过**双击**进行触发
- 添加了一个`Timeline.clear()`方法用于精准清除动画缓存
- `CommonJS`模块导出名称从`Timeline`改为`TimelineJS`
- 新增功能：支持鼠标框选并拖拽一个或多个关键帧

## 安装

#### cdn
```html
<script src="timeline/timeline.js"></script>
```


## 使用

### 1. 动画
```js
  // 注意！在这个 Fork 版本里我已经将 anim() 方法挂到了 Timeline 对象下
  // 因此不再有全局的 anim 方法，只能通过 Timeline 来访问 anim
  const { anim } = Timeline;
  anim(target).to(delay, {property:value,...}, duration, easing);
```

指定`target`后你可以链式引用任意数量的动画，如果希望多个属性动画平行进行，只需要在同一目标上再次使用`anim()`方法
#### 1.1 基本用法

- 将目标的`x`属性在`1`秒内移动到`100`，并使用`Quadratic.EaseIn`缓动，
- 停留5秒，再在`2`秒内移动回`0`

```js
anim(sprite)
  .to({x:100}, 1, Timeline.Easing.Quadratic.EaseIn)
  .to(5, {x:0}, 2);
```

#### 1.2 平行动画

同时驱动`rect`元素的宽和高，使其数值在`3`秒内变化到`50`和`75`，
并让其透明度一起变化，数值在`4`秒内变化到`0`

```js
anim(rect).to({width:50, height:75}, 3);
anim(rect).to({opacity:0}, 4);
```

### 2. 时间轴面板

```html
<script src="timeline/gui.js"></script>
<script>
  const { anim } = Timeline;
  anim(targetName, target).to(delay, {property:value,...}, duration, easing);
</script>
```

在已创建动画的情况下，添加`gui.js`会在页面底部启用**时间轴面板**，并为每个属性动画创建一条编辑轨道。在时间轴面板上可以点击和拖动关键帧对动画进行调整，在轨道空白区域上双击可以创建新的关键帧，点击**导出按钮**则可以拷贝代码；
> 注意：在此 Fork 版本中，需要双击关键帧进行编辑，并且现在支持框选多个关键帧进行拖动

在使用**时间轴面板**时，我们需要在`anim()`方法中指定动画目标的别名（这个别名也会被用于导出代码），
每个使用`to()`驱动的属性，都可以在时间轴面板中找到其轨道；

动画数据会自动记录在`localStorage`中，这样在刷新页面时动画将仍然能进行播放，被缓存的同名动画将会忽略`to()`方法，这意味着你可能即便修改了代码动画也无任何变化，但你可以手动清除掉缓存。
> 注意：在此 Fork 版本中我添加了一个`Timeline.clear()`方法用于清除已被缓存的动画


#### 2.1 时间轴示例

将`rect`元素以指定的别名`(rect)`显示在时间轴上，并让其`x`和`y`属性维持原值，
由于别名将被用于代码导出，因此强烈建议使用元素的唯一`id`作为其别名

```js
anim("rect", rect).to({x:rect.x, y:rect.y});
```


### 3. Node.js

首先从`npm`安装 `timeline-js`

`npm install timeline-js`
> 注意：npm目前下载的仍是原作者的v0.1版，建议暂时从我的repo下载文件到本地进行引用，以后我可能会重构一个版本发布到npm！

然后这样使用

```js
// 注意：Fork版将模块名称从 Timeline 改为了 TimelineJS
const TimelineJS = require('timeline-js');
const Timeline = TimelineJS.Timeline;
const { anim } = Timeline;
anim(target).to(delay, {property:value,...}, duration, easing);
```

CHANGELOG.txt
v0.2 / 2022-08-26
# 新增
- 加入了框选关键帧功能
- 加入了拖动多个关键帧功能
- 加入关键帧时间编辑

# 优化
- 界面调整
- 将原版动画方式从setInterval改为requestAnimationFrame

# 已知问题
- 这是原v0.1版就存在的：前后拖动播放头预览动画可能会出现动画丢失的情况，需要重新点击播放修正


v0.1 / 2011-05-01
- Initial version

examples/js/cssAnim2.js
(function(){
  const { log, clear } = console
  clear()

  /**
   * 核心库　http://marcinignac.com/blog/timeline-js/
   */

  // 清除缓存的动画
  // 不知道为什么它要把动画写到浏览器缓存, 已缓存的动画不会接受新的to()方法, 最好清理一下不然调试时代码改了数值也看不见
  // localStorage.setItem('timeline.js.data.Global', '')


  // 设置动画无限循环
  const tl = Timeline.getGlobalInstance()
  tl.clear()
  tl.loop(-1)
  tl.fps = 24 /* 这玩意更像是动画时间缩放,因为无论怎样动画看上去都是60fps的,只是动画速度变化了 */


  // 动画参数格式(见animProps)
  const box = document.querySelector('#box')
  const animProps = {
      element: box,
      x: 0,
      y: 0,
      angle: 0,
      scale: 0,
      opacity: 0,
      hue: 0,
      rounded: 0
  }


  // 添加关键帧, 第一个参数是轨道名, 同名轨道不会再次接受to()方法, 需要手动清理localStorage(见上)
  const tlAnim = Timeline.anim

  tlAnim('box', animProps)
    .to({'x':50}, 0)
    .to({'x':400}, 1, Timeline.Easing.Cubic.EaseIn)
    .to({'x':50}, 1, Timeline.Easing.Cubic.EaseOut)

  tlAnim('box', animProps)
    .to({'y':50}, 0)
    .to({'y':400}, 1, Timeline.Easing.Cubic.EaseIn)
    .to({'y':50}, 1, Timeline.Easing.Cubic.EaseOut)

  tlAnim('box', animProps)
    .to({'angle':0},0)
    .to({'angle':360}, 1, Timeline.Easing.Cubic.EaseIn)
    .to({'angle':720}, 1, Timeline.Easing.Cubic.EaseOut)

  tlAnim('box', animProps)
    .to({'scale':1}, 0)
    .to({'scale':.4}, 1, Timeline.Easing.Cubic.EaseIn)
    .to({'scale':1}, 1, Timeline.Easing.Cubic.EaseOut)

  tlAnim('box', animProps)
    .to({'opacity':1}, 0)
    .to({'opacity':.4}, 1, Timeline.Easing.Cubic.EaseIn)
    .to({'opacity':1}, 1, Timeline.Easing.Cubic.EaseOut)

  tlAnim('box', animProps)
    .to({'hue':1}, 0)
    .to({'hue':180}, 1, Timeline.Easing.Cubic.EaseIn)
    .to({'hue':360}, 1, Timeline.Easing.Cubic.EaseOut)

  tlAnim('box', animProps)
    .to({'rounded':0}, 0)
    .to({'rounded':50}, 1, Timeline.Easing.Cubic.EaseIn)
    .to({'rounded':0}, 1, Timeline.Easing.Cubic.EaseOut)

  // 打印数值变化
  // setInterval(`log(animProps)`, 500)


  // 自行实现动画渲染
  // timeline-js 不与动画耦合,只驱动数值,因此可用于 DOM/SVG/3D 等不同场景,主要是它还带了个时间轴面板!!
  rAF();
  function rAF() {
    animProps.element.style.left = `${animProps.x}px`
    animProps.element.style.top = `${animProps.y}px`
    animProps.element.style.opacity = `${animProps.opacity}`
    animProps.element.style.borderRadius = `${animProps.rounded}px`
    animProps.element.style.filter = `hue-rotate(${animProps.hue}deg)`
    animProps.element.style.transform = `rotate(${animProps.angle}deg) scale(${animProps.scale}) translateZ(0)`
    requestAnimationFrame(rAF, document.body)
  }


  // 按钮事件
  const btns = document.querySelectorAll('.btn')
  let timelineIsShow = true;
  btns.forEach(btn => {
      btn.addEventListener('click', e => {
          let name = e.currentTarget.className
          if (name.includes('play')) { tl.play() }
          else if (name.includes('pause')) { tl.pause() }
          else if (name.includes('stop')) { tl.stop() }
          else if (name.includes('toggle')) {
            if (timelineIsShow) {
              tl.container.style.bottom = -tl.container.offsetHeight + 'px'
            } else {
              tl.container.style.bottom = 0
            }
            timelineIsShow = !timelineIsShow
          }
      })
  })
  
}())

src/timeline.js
/**
 * Timeline.js
 * A compact JavaScript animation library with a GUI timeline for fast editing.
 * 
 * v0.2 - Forked by WaveF (wavef@qq.com) / 2022-08-26
 * v0.1 - Created by Marcin Ignac (http://marcinignac.com) / 2011-05-01
 */

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
// IN THE SOFTWARE.

(function(root){
  var Timeline = function() {
    this.name = "Global";
    this.anims = [];
    this.time = 0;
    this.totalTime = 0;
    this.loopCount = 0;
    this.loopMode = 0;
    this.playing = true;
    this.version = 0.2;
    var self = this;
    this.fps = 30;

    raf(self.update.bind(self), this.fps);
  };

  function raf(callback, fps) {
    let fpsInterval = 1000 / fps, then = now = Date.now(), elapsedTime = 0;
    let loop = () => {
      requestAnimationFrame(loop);
      now = Date.now();
      elapsedTime = now - then;
      if (elapsedTime > fpsInterval) {
        then = now - (elapsedTime % fpsInterval);
        callback();
      }
    };
    loop();
  }

  Timeline.getGlobalInstance = function() {
    if (!Timeline.globalInstance) {
      Timeline.globalInstance = new Timeline();
    }
    return Timeline.globalInstance;
  };

  //Possible values of n:
  //-1 infinite loop
  //0  play forever without looping, continue increasing time even after last animation
  //1  play once and stop at the time the last animation finishes
  //>1 loop n-times
  Timeline.prototype.loop = function(n) {
    this.loopMode = n;
  };

  Timeline.prototype.stop = function() {
    this.playing = false;
    this.time = 0;
  };

  Timeline.prototype.pause = function() {
    this.playing = false;
  };

  Timeline.prototype.play = function() {
    this.playing = true;
  };

  Timeline.prototype.preUpdate = function() {
    //placeholder for hooks like GUI rendering
  };

  Timeline.prototype.update = function(dt) {
    var deltaTime = dt || 1 / this.fps;

    this.preUpdate();

    if (this.playing) {
      this.totalTime += deltaTime;
      this.time += deltaTime;
    }

    if (this.loopMode !== 0) {
      var animationEnd = this.findAnimationEnd();
      if (this.time > animationEnd) {
        if (this.loopMode == -1 || (this.loopCount < this.loopMode)) {
          this.time = 0;
          this.loopCount++;
          for(var i=0; i<this.anims.length; i++) {
            this.anims[i].hasStarted = false;
            this.anims[i].hasEnded = false;
          }
        }
        else {
          this.playing = false;
        }
      }
    }

    this.applyValues();
  };

  Timeline.prototype.findAnimationEnd = function() {
    var endTime = 0;
    for(var i=0; i<this.anims.length; i++) {
      if (this.anims[i].endTime > endTime) {
        endTime = this.anims[i].endTime;
      }
    }
    return endTime;
  };

  Timeline.prototype.applyValues = function() {
    for(var i=0; i<this.anims.length; i++) {
      var propertyAnim = this.anims[i];
      if (this.time < propertyAnim.startTime || propertyAnim.hasEnded) {
        continue;
      }

      if (this.time >= propertyAnim.startTime && !propertyAnim.hasStarted) {
        var startValue = propertyAnim.target[propertyAnim.propertyName];
        if (startValue.length && startValue.indexOf('px') > -1) {
          propertyAnim.startValue = Number(startValue.replace('px', ''));
          propertyAnim.unit = 'px';
        }
        else {
          propertyAnim.startValue = Number(startValue);
        }
        propertyAnim.hasStarted = true;
        if (propertyAnim.onStart) {
          propertyAnim.onStart();
        }
      }
      var duration = propertyAnim.endTime - propertyAnim.startTime;
      var t = duration ? (this.time - propertyAnim.startTime)/(duration) : 1;
      t = Math.max(0, Math.min(t, 1));
      t = propertyAnim.easing(t);

      var value = propertyAnim.startValue + (propertyAnim.endValue - propertyAnim.startValue) * t;
      // console.log(value)

      if (propertyAnim.unit) value += propertyAnim.unit;
      propertyAnim.target[propertyAnim.propertyName] = value;

      if (propertyAnim.parent && propertyAnim.parent.onUpdateCallback) {
        propertyAnim.parent.onUpdateCallback(propertyAnim);
      }

      if (this.time >= propertyAnim.endTime && !propertyAnim.hasEnded) {
        propertyAnim.hasEnded = true;
        if (propertyAnim.onEnd) {
          propertyAnim.onEnd();
        }
      }

      if (t == 1) {
        if (this.loopMode == 0) {
          this.anims.splice(i, 1);
          i--;
        }
      }
    }
  };

  //--------------------------------------------------------------------

  function Anim(name, target, timeline) {
    this.startTime = 0;
    this.endTime = 0;
    this.time = 0;
    this.propertyAnims = [];
    this.hasStarted = false;
    this.hasEnded = false;

    this.name = name;
    this.target = target;
    this.timeline = timeline;
    this.animGroups = [];
  }

  //delay, properties, duration, easing
  Anim.prototype.to = function() {
    var args = [];
    for(var i=0; i<arguments.length; i++) {
      args.push(arguments[i]);
    }
    var delay;
    var properties;
    var duration;
    var easing;

    if (typeof(args[0]) == "number") {
      delay = args.shift();
    }
    else {
      delay = 0;
    }

    if (typeof(args[0]) == "object") {
      properties = args.shift();
    }
    else {
      properties = {};
    }

    if (typeof(args[0]) == "number") {
      duration = args.shift();
    }
    else {
      duration = 1;
    }

    if (typeof(args[0]) == "function") {
      easing = args.shift();
    }
    else {
      easing = Timeline.Easing.Linear.EaseNone;
    }

    var animGroup = [];
    var nop = function() {}

    for(var propertyName in properties) {
      var animInfo = {
        hasStarted: false,
        timeline: this.timeline,
        targetName: this.name,
        target: this.target,
        propertyName: propertyName,
        endValue: properties[propertyName],
        delay: delay,
        startTime: this.timeline.time + delay + this.endTime,
        endTime: this.timeline.time + delay + this.endTime + duration,
        easing: easing,
        parent: this,
        onStart: nop,
        onEnd: nop
      };
      this.timeline.anims.push(animInfo);
      animGroup.push(animInfo);
    }
    this.animGroups.push(animGroup);
    this.endTime += delay + duration;
    return this;
  };

  Anim.prototype.onStart = function(callback) {
    var currentAnimGroup = this.animGroups[this.animGroups.length-1];
    if (!currentAnimGroup) return;

    var called = false;

    currentAnimGroup.forEach(function(anim) {
      anim.onStart = function() {
        if (!called) {
          called = true;
          callback();
        }
      };
    })

    return this;
  }

  Anim.prototype.onUpdate = function(callback) {
    var self = this;
    this.onUpdateCallback = function() {
      callback();
    };
    return this;
  }

  Anim.prototype.onEnd = function(callback) {
    var currentAnimGroup = this.animGroups[this.animGroups.length-1];
    if (!currentAnimGroup) return;

    var called = false;

    currentAnimGroup.forEach(function(anim) {
      anim.onEnd = function() {
        if (!called) {
          called = true;
          callback();
        }
      }
    })

    return this;
  }

  function anim(targetName, targetObject, parentTimeline) {
    var args = [];
    for(var i=0; i<arguments.length; i++) {
      args.push(arguments[i]);
    }
    var name;
    var target;
    var timeline;

    if (typeof(args[0]) == "string") {
      name = args.shift();
    }

    if (typeof(args[0]) == "object") {
      target = args.shift();
    }
    else {
      target = {};
    }

    if (typeof(args[0]) == "object") {
      timeline = args.shift();
    }
    else {
      timeline = Timeline.getGlobalInstance();
    }

    return new Anim(name, target, timeline);
  }

  //--------------------------------------------------------------------

  Timeline.Easing = { Linear: {}, Quadratic: {}, Cubic: {}, Quartic: {}, Quintic: {}, Sinusoidal: {}, Exponential: {}, Circular: {}, Elastic: {}, Back: {}, Bounce: {} };

  Timeline.Easing.Linear.EaseNone = function ( k ) {
    return k;
  };

  Timeline.Easing.Quadratic.EaseIn = function ( k ) {
    return k * k;
  };

  Timeline.Easing.Quadratic.EaseOut = function ( k ) {
    return - k * ( k - 2 );
  };

  Timeline.Easing.Quadratic.EaseInOut = function ( k ) {
    if ( ( k *= 2 ) < 1 ) return 0.5 * k * k;
    return - 0.5 * ( --k * ( k - 2 ) - 1 );
  };

  Timeline.Easing.Cubic.EaseIn = function ( k ) {
    return k * k * k;
  };

  Timeline.Easing.Cubic.EaseOut = function ( k ) {
    return --k * k * k + 1;
  };

  Timeline.Easing.Cubic.EaseInOut = function ( k ) {
    if ( ( k *= 2 ) < 1 ) return 0.5 * k * k * k;
    return 0.5 * ( ( k -= 2 ) * k * k + 2 );
  };

  Timeline.Easing.Elastic.EaseIn = function( k ) {
    var s, a = 0.1, p = 0.4;
    if ( k === 0 ) return 0; if ( k == 1 ) return 1; if ( !p ) p = 0.3;
    if ( !a || a < 1 ) { a = 1; s = p / 4; }
    else s = p / ( 2 * Math.PI ) * Math.asin( 1 / a );
    return - ( a * Math.pow( 2, 10 * ( k -= 1 ) ) * Math.sin( ( k - s ) * ( 2 * Math.PI ) / p ) );
  };

  Timeline.Easing.Elastic.EaseOut = function( k ) {
    var s, a = 0.1, p = 0.4;
    if ( k === 0 ) return 0; if ( k == 1 ) return 1; if ( !p ) p = 0.3;
    if ( !a || a < 1 ) { a = 1; s = p / 4; }
    else s = p / ( 2 * Math.PI ) * Math.asin( 1 / a );
    return ( a * Math.pow( 2, - 10 * k) * Math.sin( ( k - s ) * ( 2 * Math.PI ) / p ) + 1 );
  };

  Timeline.Easing.Elastic.EaseInOut = function( k ) {
    var s, a = 0.1, p = 0.4;
    if ( k === 0 ) return 0; if ( k == 1 ) return 1; if ( !p ) p = 0.3;
    if ( !a || a < 1 ) { a = 1; s = p / 4; }
    else s = p / ( 2 * Math.PI ) * Math.asin( 1 / a );
    if ( ( k *= 2 ) < 1 ) return - 0.5 * ( a * Math.pow( 2, 10 * ( k -= 1 ) ) * Math.sin( ( k - s ) * ( 2 * Math.PI ) / p ) );
    return a * Math.pow( 2, -10 * ( k -= 1 ) ) * Math.sin( ( k - s ) * ( 2 * Math.PI ) / p ) * 0.5 + 1;
  };

  Timeline.Easing.Back.EaseIn = function( k ) {
    var s = 1.70158;
    return k * k * ( ( s + 1 ) * k - s );
  };

  Timeline.Easing.Back.EaseOut = function( k ) {
    var s = 1.70158;
    return ( k = k - 1 ) * k * ( ( s + 1 ) * k + s ) + 1;
  };

  Timeline.Easing.Back.EaseInOut = function( k ) {
    var s = 1.70158 * 1.525;
    if ( ( k *= 2 ) < 1 ) return 0.5 * ( k * k * ( ( s + 1 ) * k - s ) );
    return 0.5 * ( ( k -= 2 ) * k * ( ( s + 1 ) * k + s ) + 2 );
  };

  Timeline.Easing.Bounce.EaseIn = function( k ) {
    return 1 - Timeline.Easing.Bounce.EaseOut( 1 - k );
  };

  Timeline.Easing.Bounce.EaseOut = function( k ) {
    if ( ( k /= 1 ) < ( 1 / 2.75 ) ) {
      return 7.5625 * k * k;
    } else if ( k < ( 2 / 2.75 ) ) {
      return 7.5625 * ( k -= ( 1.5 / 2.75 ) ) * k + 0.75;
    } else if ( k < ( 2.5 / 2.75 ) ) {
      return 7.5625 * ( k -= ( 2.25 / 2.75 ) ) * k + 0.9375;
    } else {
      return 7.5625 * ( k -= ( 2.625 / 2.75 ) ) * k + 0.984375;
    }
  };

  Timeline.Easing.Bounce.EaseInOut = function( k ) {
    if ( k < 0.5 ) return Timeline.Easing.Bounce.EaseIn( k * 2 ) * 0.5;
    return Timeline.Easing.Bounce.EaseOut( k * 2 - 1 ) * 0.5 + 0.5;
  };

  Timeline.easingFunctionToString = function( f ) {
    for(var name in Timeline.easingMap) {
      if (Timeline.easingMap[name] == f) {
        return name;
      }
    }
  };

  Timeline.stringToEasingFunction = function( name ) {
    return Timeline.easingMap[name];
  };

  Timeline.easingMap = {
  };

  for(var easingFunctionFamilyName in Timeline.Easing) {
    var easingFunctionFamily = Timeline.Easing[easingFunctionFamilyName];
    for(var easingFunctionName in easingFunctionFamily) {
      Timeline.easingMap[easingFunctionFamilyName + "." + easingFunctionName] = easingFunctionFamily[easingFunctionName];
    }
  }


  //--------------------------------------------------------------------

  Timeline.anim = anim;

  if (typeof define === "function" && define.amd) {
      /* AMD */
      define('Timeline', Timeline);
  } else if (typeof module === "object" && typeof exports === "object") {
      /* CommonJS */
      module.exports['TimelineJS'] = Timeline;
  } else if (typeof exports === "object") {
      /* CommonJS2 */
      exports['TimelineJS'] = Timeline;
  } else {
      /* ES5 */
      root['Timeline'] = Timeline;
  }

}(this));